gcd/src/main/java/org/spine3/server/storage/datastore/DsAggregateStorage.java

```diff
@@ -89,34 +95,46 @@ public void writeEventCountAfterLastSnapshot(I id, int eventCount) {
     }
 
     @Override
-    protected void writeInternal(I id, AggregateStorageRecord record) {
+    protected void writeRecord(I id, AggregateStorageRecord record) {
         checkNotNull(id);
 
-        final Value.Builder idValue = makeValue(idToString(id));
-        final Property.Builder idProperty = makeProperty(AGGREGATE_ID_PROPERTY_NAME, idValue);
-        final Entity.Builder entity = messageToEntity(record, makeKey(KIND));
-        entity.addProperty(idProperty);
-        entity.addProperty(DatastoreProperties.makeTimestampProperty(record.getTimestamp()));
-        entity.addProperty(DatastoreProperties.makeTimestampNanosProperty(record.getTimestamp()));
-
-        final Mutation.Builder mutation = Mutation.newBuilder().addInsertAutoId(entity);
-        datastore.commit(mutation);
+        final String stringId = idToString(id);
+        final KeyFactory keyFactory = datastore.getKeyFactory(KIND);
+        String eventId = record.getEventId();
+        if (eventId.isEmpty()) {
+            // Snapshots have no Event IDs.
+            eventId = SNAPSHOT + stringId;
+        }
+        final Key key = keyFactory.newKey(eventId);
+        final Entity incompleteEntity = Entities.messageToEntity(record, key);
+        final Entity.Builder builder = Entity.builder(incompleteEntity);
+        builder.set(AGGREGATE_ID_PROPERTY_NAME, stringId);
+        datastore.createOrUpdate(builder.build());
     }
 
     @Override
     protected Iterator<AggregateStorageRecord> historyBackward(I id) {
         checkNotNull(id);
 
         final String idString = idToString(id);
-        final Filter.Builder idFilter = makeFilter(AGGREGATE_ID_PROPERTY_NAME, EQUAL,
-                makeValue(idString));
-        final Query.Builder query = DatastoreQueries.makeQuery(DESCENDING, KIND);
-        query.setFilter(idFilter).build();
-
-        final List<EntityResult> entityResults = datastore.runQuery(query);
-        final List<AggregateStorageRecord> records = entitiesToMessages(entityResults, TYPE_URL);
-        final Iterator<AggregateStorageRecord> recordsIterator = records.iterator();
-        return recordsIterator;
+        final Query<?> query = Query.entityQueryBuilder()
```

Comment:
Use chain-of-calls code conventions of spine. In common there are lot of such situation. You can use format settings from spine core to find them all.